### PR TITLE
[HYBRID FRONTEND] Modify hybrid script to new interface; hybrid op supported; enable compilation_database in CMakeList.txt

### PR DESCRIPTION
**Pull request by were**
Move old hybrid script frontend
```` Python
import tvm
@tvm.hybrid.script
def outer_product(a, b, c):
    for j in range(128):
        for i in range(128):
            c[i,j] = a[i] * b[j]

a = tvm.placeholder((128, ), 'float32', 'a')
b = tvm.placeholder((128, ), 'float32', 'b')
c = tvm.placeholder((128, ), 'float32', 'b')
# returns c's op
# if it is multi-output, return the tuple of output ops
op = outer_product(a, b, c)
````
to
````Python
import tvm
@tvm.hybrid.script
def outer_product(a, b):
    c = tvm.output((128, 128), dtype)
    for j in range(128):
        for i in range(128):
            c[i,j] = a[i] * b[j]
    return c

a = tvm.placeholder((128, ), 'float32', 'a')
b = tvm.placeholder((128, ), 'float32', 'b')
# returns an ExternOp
c = outer_product(a, b) 
````

### CODE REVIEW

**Comment by tqchen** [comment]
Ideally, we should keep the old function parse as it is. And instead support a __call__ of the hybrid script that calls parse and then generate an ExternOp.

**Comment by were** [comment]
OK. Then I know how an op works.

**Comment by tqchen** [comment]
## Multi output case
```python
import tvm
@tvm.hybrid.script
def outer_product(a, b):
    c = tvm.output((128, 128), dtype)
    d = tvm.output((128, 128), dtype)

    for j in range(128):
        for i in range(128):
            c[i,j] = a[i] * b[j]
            d[i,j] = a[i] + 1
    return c, d

a = tvm.placeholder((128, ), 'float32', 'a')
b = tvm.placeholder((128, ), 'float32', 'b')
# returns an ExternOp
c0, c1 = outer_product(a, b) 
d = tvm.compute(c.shape, lambda *i: c0(*i)+c1(*i))
sch = tvm.create_schedule(d.op)
.
```

## Inplace Update
I agree that is harder to express an inplace update. The general design goal of the operator relation is that everything is immutable and there is no inplace update. We can still assign c and a to be the same buffer to achieve the effect of of inplace update though.

**Comment by were** [comment]
I like the old one better for 2 reasons:
1. Calling a hybrid script becomes more like calling a TVM built module, which makes code clean when doing sw-emulation verification
2. In-place algorithm support will be rollback, if we use the new style
Why don't we do this?
```
import tvm
@tvm.hybrid.script
def outer_product(a, b, c):
    for j in range(128):
        for i in range(128):
            c[i,j] = a[i] * b[j]

a = tvm.placeholder((128, ), 'float32', 'a')
b = tvm.placeholder((128, ), 'float32', 'b')
c = tvm.placeholder((128, ), 'float32', 'b')
# returns c's op
# if it is multi-output, return the tuple of output ops
op = outer_product(a, b, c)
d = tvm.compute(c.shape, lambda *i: c(*i)+1)
sch = tvm.create_schedule(d.op)
```

**Comment by tqchen** [comment]
@were can you update this PR and add a few examples?

**Comment by were** [comment]
@tqchen @kevinthesun 
I am facing a problem like this:
````Python
op = _ExternOp(input_placeholders, input_buffers, output_buffers, ..., body)
ret = [op.output(i) for i in range(len(output_buffers))]
return ret[0] if len(ret) == 1 else ret
````
Originally, what I am supposed to do is to register an op and return its output tensors, which seems to be OK, but the problem I am facing is that all the output tensors' names are same as the op's name.

Just as it is mentioned above, `return c, d` gives different names to the output tensors.

Also, if I want to give a different name to the tensor, how can I specify the relationship between the tensor and op?

**Comment by tqchen** [comment]
The output tensor's name can be viewed as op_name.value0, opname.value1 which have different meaning. Note that under the hood tvm uses pointer uniqueness, so even if the name are the same it is fine

**Comment by were** [comment]
OK. I know know what you mean --- no matter how users name the output tensors, we will finally convert all the tensors to name `foo.v0`, `foo.v1`, and etc.

Then a new problem comes when doing in-place. (or more accurately, it is not a problem. I just want to make sure it is acceptable):
````Python
out = foo(a)
sch = tvm.create_schedule(a.op)
tvm.build(sch, [out])
````
Say, `foo` is an in-place function call which transforms `a` to `out`. They are essentially the same buffer with alias, but I CANNOT use `a` when compilation. I need to force users to pass `out`, or the compiler cannot find the buffer in the binds.

**Comment by tqchen** [comment]
For now, we could avoid supporting inplace functions, if we want to create an inplace op, we can do 
```python
out = foo(a)
sch = tvm.create_schedule(a.op)
buffer = tvm.decl_buffer(shape=out.shape)
tvm.build(sch, [buffer], binds={a:buffer, out:buffer})
```

**Comment by were** [comment]
I am refusing not to support in-place because it is some kind of "rollback" --- it is supported before but disabled just because the "glue" to the framework.

**Comment by tqchen** [comment]
The example I gave kinda of support inplace operations

**Comment by were** [comment]
For now, I think it is good enough.
My question is that  we have already have buffer binds stored in Op's attributes, so can we do it spontaneously?

**Comment by tqchen** [comment]
It sounds good to me, as long as the behavior is well documented

**Comment by were** [comment]
@tqchen @kevinthesun 
Sorry, I was busy for too much my own stuff these two months. Finally, I got a little bandwidth so that I could return here. The workloads to refactor the interfaces and glue it to an `OpNode` are far beyond my expectation.
I am struggling for two missing pieces:
1. After putting a lowered body to an `ExternOp`, I face a paradox: the `ExternOp.output(i)` will return a tensor which points to the `ExternOp` itself, which means this tensor has the same name as the `ExternOp`. However, this tensor has its own name defined by the users `a=output(shape, dtype)`.
I can either:
    - Add a rename pass to make it compatible with `ExternOp` interface
    - Add my own `HybridOp` and overwrite `.output(i)` member.
2. The in-place thing, actually, it is redundant to return this tensor again in the output list. The contribution of the function has been delivered implicitly by the side effect. My suggestion is that we just put it in the input list, and we do not put it in the output list. ** I believe TVM has no global (inter-op) check which leverages the input/output information, which is based on side-effect free assumption? **

If 2.'s bold assumption is not true or not acceptable for further extension of this project, I will pick the second solution in 1.

**Comment by were** [comment]
@tqchen If CI test passed, I believe this PR is ready for review.

**Comment by were** [comment]
@kevinthesun @xqdan @tqchen Can you guys take a look so that I can get it in?

**Comment by xqdan** [comment]
Can we update this place? Since we have an unified routine after this patch.
https://github.com/dmlc/tvm/blob/master/python/tvm/build_module.py#L343

File: python/tvm/hybrid/parser.py [hunk 1]
```diff
@@ -72,15 +73,17 @@ def __init__(self, args, usage, func_name=None):
...
         self.usage = usage.copy()
         self._args = {} # Dict maps arg name to actual arg instance (either a var or a buffer)
-        self.var_buffers = {} # Buffers formed by mutatble variables
         self.alloc_buffers = {} # Buffers formed by allocate instructions
         self.loops_above = {} # State variable that indicates loop levels above the current node
         self.var_consts = {} # Variables that are determined as readonly in previous stage
         self.func_name = func_name # The name of the function to be lowered
-        self.iter_axis = []
+        self.outputs = [] # Output tensors' name
+        self.side_effect = set() # Tensors with side effects
```
Comment (xqdan) [inline python/tvm/hybrid/parser.py]: Just a question, what's side_effect for?

**Comment by xqdan** [comment]
> **I believe TVM has no global (inter-op) check which leverages the input/output information based on side-effect free assumption.**
> 
> If 2.'s bold assumption is not true or not acceptable for further extension of this project, I will pick the second solution in 1.

I think when we need to support inline/fuse between ComputeOp and HybirdOp, the data dependence between the Ops is important.

File: python/tvm/hybrid/parser.py (hunk 1 above)
Comment (were) [reply python/tvm/hybrid/parser.py]: Later for in-place operation.

**Comment by were** [comment]
@xqdan Actually, current hybrid op is the same as the external op, which cannot be splitted. Thus, I think no inline will be done?

File: src/op/hybrid_op.cc [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    bind_spec.push_back(buffer);
+    bind_spec.push_back(tensor);
+    for (size_t k = 0; k < buffer->shape.size(); ++k) {
+      tuple.push_back(make_const(buffer->shape[k].type(), 0));
+      tuple.push_back(buffer->shape[k]);
+    }
+    ret = AttrStmt::make(
+        bind_spec, attr::buffer_bind_scope,
+        Call::make(Handle(), intrinsic::tvm_tuple, tuple, Call::Intrinsic), ret);
+  };
```
Comment (xqdan) [inline src/op/hybrid_op.cc]: please put comments about why we need binding here

File: src/op/op_util.cc [hunk 3]
```diff
@@ -164,6 +164,43 @@ std::vector<Stmt> MakeIfNest(const std::vector<Expr>& predicates) {
...
+  const std::unordered_map<Tensor, Tensor>& vmap_;
+};
+
+Stmt ReplaceProvideTensor(Stmt stmt,
+                   const std::unordered_map<Tensor, Tensor>& replace) {
+  ProviderReplacer repl(replace);
+  Stmt ret = repl.Mutate(stmt);
+  return repl.found ? ret : stmt;
+}
+Expr ReplaceProvideTensor(Expr expr,
```
Comment (xqdan) [inline src/op/op_util.cc]: Only for Stmt is enough

Comment (were) [reply src/op/op_util.cc]: Jesus, let me fix it!

File: src/op/hybrid_op.cc [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    f_push_bind(buffer, inputs[i]);
+  }
+
+  std::unordered_map<Tensor, Tensor> rmap;
+  for (int i = 0; i < this->num_outputs(); ++i) {
+    rmap[outputs[i]] = stage->op.output(i);
+  }
+  auto n = make_node<HybridOpNode>(*this);
+  ret = op::ReplaceTensor(ret, rmap);
+  ret = op::ReplaceProvideTensor(ret, rmap);
```
[the remark this answers is not in the document]
Comment (were) [reply src/op/hybrid_op.cc]: OK...

File: src/op/hybrid_op.cc (hunk 2 above)
Comment (were) [reply src/op/hybrid_op.cc]: this is just copy paste from extern_op.cc

**Comment by were** [comment]
@xqdan That's why I opened a new op node. Later more features will be supported and more attributes will be added.

File: src/op/hybrid_op.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinthesun) [reply src/op/hybrid_op.cc]: Just curious, "In hybrid script all the tensors have their own names specified by the users." Does this mean that user can write something like "out = my_out_tensor((1, 3, 224, 224), 'float')"? In this case, if out is generated by, say flatten op with name 'tensor', out has different name with flatten op. This is why we need replace tensors?

**Comment by kevinthesun** [comment]
Update tutorial as well?

File: src/op/hybrid_op.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (were) [reply src/op/hybrid_op.cc]: Yes. Actually, you can dump both `ret` in cpp and `op.body` in python to see this.

[the remark this answers is not in the document]
Comment (kevinthesun) [reply src/op/hybrid_op.cc]: Got it. Maybe we can document this behavior somewhere, such as tvm.hybrid.script,  make it transparent to user.

**Comment by tqchen** [comment]
Thanks @xqdan @kevinthesun for reviews, Thanks @were this is merged!

**Pull request merged**